static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -0,0 +1,202 @@
+  const hasInvalidBranchSequence = (tag.key.match(/\.{2,}/g) ?? []).length > 0;
+
+  // Ignore tags with 0, or >4 branches, as well as sequential dots (e.g. 'some..tag')
+  if (hasInvalidBranchCount || hasInvalidBranchSequence) {
+    tree[tag.key] = {value: tag.value, subtree: {}, meta, originalTag};
+    return tree;
+  }
+  // E.g. 'device.model.version'
+  const splitIndex = tag.key.indexOf('.'); // 6
+  const trunk = tag.key.slice(0, splitIndex); // 'device'
+  const branch = tag.key.slice(splitIndex + 1); // 'model.version'
+
+  if (tree[trunk] === undefined) {
+    // We need to define the default as an empty space for alignment
+    tree[trunk] = {value: ' ', subtree: {}};
+  }
+  // Recurse with a pseudo tag, e.g. 'model', to create nesting structure
+  const pseudoTag = {
+    key: branch,
+    value: tag.value,
+  };
+  tree[trunk].subtree = addToTagTree(tree[trunk].subtree, pseudoTag, meta, originalTag);
+  return tree;
+}
+
+interface TagsTreeKeysProps {
+  content: TagTreeContent;
+  tag: string;
+}
+
+function TagTreeKeys({content, tag}: TagsTreeKeysProps) {
+  const subtreeTags = Object.keys(content.subtree);
+  return (
+    <TreeKeyTrunk>
+      <TreeKey>{tag}</TreeKey>
+      {subtreeTags.map((t, i) => (
+        <TreeSubtreeWrapper key={`${tag}-${i}-key`} isLast={i === subtreeTags.length - 1}>
+          <TreeSpacer />
+          <TagTreeKeys tag={t} content={content.subtree[t]} />
+        </TreeSubtreeWrapper>
+      ))}
+    </TreeKeyTrunk>
+  );
+}
+interface TagsTreeValuesProps extends TagsTreeKeysProps {
+  projectId: string;
+  projectSlug: string;
+  streamPath: string;
+}
+
+function TagTreeValues({content, tag, ...props}: TagsTreeValuesProps) {
+  const subtreeTags = Object.keys(content.subtree);
+  const organization = useOrganization();
+  const location = useLocation();
+  const originalTag = content.originalTag;
+  return (
+    <TreeValueTrunk>
+      <TreeValue>
+        {originalTag ? (
+          <EventTagsContent
+            tag={originalTag}
+            organization={organization}
+            query={generateQueryWithTag(
+              {...location.query, referrer: 'event-tags-tree'},
+              originalTag
+            )}
+            meta={content?.meta ?? {}}
+            {...props}
+          />
+        ) : (
+          content.value
+        )}
+      </TreeValue>
+      {subtreeTags.map((t, i) => (
+        <TagTreeValues
+          key={`${tag}-${i}-value`}
+          tag={t}
+          content={content.subtree[t]}
+          {...props}
+        />
+      ))}
+    </TreeValueTrunk>
+  );
+}
+
+interface EventTagsTreeProps {
+  projectId: string;
+  projectSlug: string;
+  streamPath: string;
+  tags: EventTag[];
+  meta?: Record<any, any>;
+}
+
+function EventTagsTree({tags, meta, ...props}: EventTagsTreeProps) {
+  const tagTree = tags.reduce<TagTree>(
+    (tree, tag, i) => addToTagTree(tree, tag, meta?.[i], tag),
+    {}
+  );
```

Comment:
i think it makes sense to useMemo this since it's somewhat expensive, it also might help to call `Object.keys(tagTree)` inside the useMemo so the result is the array you will map over
```suggestion
  const tagTree = tags.reduce<TagTree>(
    (tree, tag, i) => addToTagTree(tree, tag, meta?.[i], tag),
    {}
  );
```